Cannot get a connection, pool error Timeout waiting for idle object

We have an error related to jdbc connection during `pio import`
It seams to occur when the number of stage in spark  is more than max size of jdbc connection pool.
for instance,  14 stages > max size is 8 (default max size of scalikejdbc is 8)

```
/opt/pio/bin/pio import --appid 1 --channel 2 --input /tmp/events.json -- $SPARK_IMPORT_ARG
[INFO] [Runner$] Submission command: /opt/spark//bin/spark-submit --conf spark.executor.extraClassPath=/usr/share/java/mysql-connector-java.jar --driver-class-path /usr/share/java/mysql-connector-java.jar --jars /usr/share/java/mysql-connector-java.jar --master local[*] --class io.prediction.tools.imprt.FileToEvents --files file:/opt/pio/conf/log4j.properties --driver-class-path /usr/share/java/mysql-connector-java.jar:/opt/pio/conf file:/opt/pio/lib/pio-assembly-0.9.5.jar --appid 1 --input /tmp/events.json --channel 2 --env PIO_ENV_LOADED=1,PIO_STORAGE_SOURCES_MYSQL_PASSWORD=pio,PIO_STORAGE_REPOSITORIES_METADATA_NAME=pio_meta,PIO_FS_BASEDIR=/opt/pio/.pio_store,PIO_STORAGE_SOURCES_MYSQL_URL=jdbc:mysql://gateway001/pio,PIO_HOME=/opt/pio,PIO_FS_ENGINESDIR=/opt/pio/.pio_store/engines,PIO_STORAGE_SOURCES_MYSQL_TYPE=jdbc,PIO_STORAGE_REPOSITORIES_METADATA_SOURCE=MYSQL,PIO_STORAGE_REPOSITORIES_MODELDATA_SOURCE=MYSQL,PIO_STORAGE_REPOSITORIES_EVENTDATA_NAME=pio_event,PIO_STORAGE_SOURCES_MYSQL_USERNAME=pio,PIO_FS_TMPDIR=/opt/pio/.pio_store/tmp,PIO_STORAGE_REPOSITORIES_MODELDATA_NAME=pio_model,PIO_STORAGE_REPOSITORIES_EVENTDATA_SOURCE=MYSQL,PIO_CONF_DIR=/opt/pio/conf
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/data0/opt/spark-1.5.1-bin-hadoop2.6/lib/spark-assembly-1.5.1-hadoop2.6.0.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/data0/opt/cloudera/parcels/CDH-5.4.7-1.cdh5.4.7.p0.3/jars/slf4j-log4j12-1.7.5.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Log4jLoggerFactory]
[INFO] [Remoting] Starting remoting
[INFO] [Remoting] Remoting started; listening on addresses :[akka.tcp://sparkDriver@192.168.6.3:50692]
[WARN] [AbstractLifeCycle] FAILED SelectChannelConnector@0.0.0.0:4040: java.net.BindException: Address already in use
[WARN] [AbstractLifeCycle] FAILED org.spark-project.jetty.server.Server@6993baa9: java.net.BindException: Address already in use
[WARN] [Utils] Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
[WARN] [AbstractLifeCycle] FAILED SelectChannelConnector@0.0.0.0:4041: java.net.BindException: Address already in use
[WARN] [AbstractLifeCycle] FAILED org.spark-project.jetty.server.Server@3d76f7d3: java.net.BindException: Address already in use
[WARN] [Utils] Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
[WARN] [MetricsSystem] Using default name DAGScheduler for source because spark.app.id is not set.
[Stage 0:>                         (0 + 14) / 14][ERROR] [Executor] Exception in task 0.0 in stage 0.0 (TID 0)
[ERROR] [Executor] Exception in task 3.0 in stage 0.0 (TID 3)
[ERROR] [Executor] Exception in task 12.0 in stage 0.0 (TID 12)
[ERROR] [Executor] Exception in task 11.0 in stage 0.0 (TID 11)
[ERROR] [Executor] Exception in task 9.0 in stage 0.0 (TID 9)
[ERROR] [Executor] Exception in task 5.0 in stage 0.0 (TID 5)
[WARN] [TaskSetManager] Lost task 5.0 in stage 0.0 (TID 5, localhost): org.apache.commons.dbcp.SQLNestedException: Cannot get a connection, pool error Timeout waiting for idle object
	at org.apache.commons.dbcp.PoolingDataSource.getConnection(PoolingDataSource.java:114)
	at scalikejdbc.CommonsConnectionPool.borrow(CommonsConnectionPool.scala:43)
	at scalikejdbc.DB$.localTx(DB.scala:257)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:141)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:124)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.NoSuchElementException: Timeout waiting for idle object
	at org.apache.commons.pool.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:1134)
	at org.apache.commons.dbcp.PoolingDataSource.getConnection(PoolingDataSource.java:106)
	... 14 more

[ERROR] [TaskSetManager] Task 5 in stage 0.0 failed 1 times; aborting job
Exception in thread "main" org.apache.spark.SparkException: Job aborted due to stage failure: Task 5 in stage 0.0 failed 1 times, most recent failure: Lost task 5.0 in stage 0.0 (TID 5, localhost): org.apache.commons.dbcp.SQLNestedException: Cannot get a connection, pool error Timeout waiting for idle object
	at org.apache.commons.dbcp.PoolingDataSource.getConnection(PoolingDataSource.java:114)
	at scalikejdbc.CommonsConnectionPool.borrow(CommonsConnectionPool.scala:43)
	at scalikejdbc.DB$.localTx(DB.scala:257)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:141)
	at io.prediction.data.storage.jdbc.JDBCPEvents$$anonfun$write$1.apply(JDBCPEvents.scala:124)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.rdd.RDD$$anonfun$foreachPartition$1$$anonfun$apply$29.apply(RDD.scala:898)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.SparkContext$$anonfun$runJob$5.apply(SparkContext.scala:1848)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:66)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.NoSuchElementException: Timeout waiting for idle object
	at org.apache.commons.pool.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:1134)
	at org.apache.commons.dbcp.PoolingDataSource.getConnection(PoolingDataSource.java:106)
	... 14 more
```

I found that there is no way to set max size of connection pool for scalikejdbc.

https://github.com/PredictionIO/PredictionIO/blob/develop/data/src/main/scala/io/prediction/data/storage/jdbc/StorageClient.scala#L39-L42

--- a/data/src/main/scala/io/prediction/data/storage/jdbc/StorageClient.scala
+++ b/data/src/main/scala/io/prediction/data/storage/jdbc/StorageClient.scala
@@ -36,10 +36,15 @@ class StorageClient(val config: StorageClientConfig)
     throw new StorageClientException("The PASSWORD variable is not set!", null)
   }
 
+  // set max size of connection pool
+  val maxSize: Int = config.properties.getOrElse("CONNECTIONS", "8").toInt
+  val settings = ConnectionPoolSettings(maxSize = maxSize)
+
   ConnectionPool.singleton(
     config.properties("URL"),
     config.properties("USERNAME"),
-    config.properties("PASSWORD"))
+    config.properties("PASSWORD"),
+    settings)
   /** JDBC connection URL. Connections are managed by ScalikeJDBC. */
   val client = config.properties("URL")
 }

--- a/docs/manual/source/system/anotherdatastore.html.md
+++ b/docs/manual/source/system/anotherdatastore.html.md
@@ -192,6 +192,11 @@ When `TYPE` is set to `jdbc`, the following configuration keys are supported.
     use when it reads from the JDBC connection, e.g.
     `PIO_STORAGE_SOURCES_PGSQL_PARTITIONS=4`
 
+-   CONNECTIONS (optional, default to 8)
+
+    This value is used by scalikejdbc library to determine the max size of connection pool, e.g.
+    `PIO_STORAGE_SOURCES_PGSQL_CONNECTIONS=8`
+
 -   INDEX (optional since v0.9.6, default to disabled)
 
     This value is used by creating indexes on entityId and entityType columns to

